### PR TITLE
Fix Cohere Store runtime loading and add quantization options

### LINKED ISSUE #339: Benchmark Cohere Transcribe GGUF quantizations on Windows
## Summary

Compare the four Cohere Transcribe 03-2026 GGUF quantizations exposed by the
Windows plugin:

- Q4_K
- Q5_0
- Q6_K
- Q8_0

The goal is to make the storage, memory, accuracy, and throughput trade-offs
reproducible before changing any product recommendation.

On the tested RTX 4060 Ti system, the larger files did **not** produce a
monotonic accuracy improvement:

- Q4_K used the least disk and process memory and had the lowest non-CJK macro
  WER in this run.
- Q5_0 had the highest aggregate throughput, but only by about 1.3% over Q4_K.
- Q6_K and Q8_0 had the lowest CJK macro CER, only 0.03 percentage points below
  Q4_K.
- Q8_0 used 0.85 GiB more peak process working set than Q4_K without improving
  non-CJK macro WER or throughput.
- All four variants completed every clip and returned an empty result for all
  silence/noise inputs.

Related implementation: #338

## Recommendation

Keep all four variants separately selectable and downloadable.

- **Q4_K is the efficiency choice** on the current evidence: smallest download,
  lowest peak working set, and no aggregate quality penalty in this run.
- **Q5_0 remains a reasonable conservative default/reference point** and was
  marginally fastest here.
- **Q6_K and Q8_0 are useful opt-in comparison choices**, but this benchmark
  does not show a consistent quality benefit that offsets their larger
  footprint.

This single-machine FLEURS run is not sufficient on its own to change the
product default. Additional CPU, Vulkan, lower-VRAM CUDA, and real microphone
dictation results are welcome in this issue.

### Environment

| Item | Value |
|---|---|
| Host app | TypeWhisper 1.0.6.0 from Microsoft Store |
| Cohere plugin | 1.0.1 build from #338 |
| TypeWhisper benchmark source | [`5e69ab6d4ec4b7c71c625f2c17bb062f7f408852`](https://github.com/TypeWhisper/typewhisper-win/tree/5e69ab6d4ec4b7c71c625f2c17bb062f7f408852) |
| Runner | [`eng/benchmark_cohere_quantizations.py`](https://github.com/TypeWhisper/typewhisper-win/blob/5e69ab6d4ec4b7c71c625f2c17bb062f7f408852/eng/benchmark_cohere_quantizations.py) |
| Runtime | CrispASR 0.8.24, CUDA |
| CPU | AMD Ryzen 7 7800X3D, 8 cores / 16 logical processors |
| Inference threads | 8 |
| GPU | NVIDIA GeForce RTX 4060 Ti, 16,380 MiB |
| NVIDIA driver | 610.88 |
| System memory | 32 GB |
| OS | Windows 11 Pro, build 26200 |
| Python | 3.10.11 |

### Pinned inputs

| Input | Revision / SHA-256 |
|---|---|
| Cohere GGUF model revision | `2242638d5dfecc6f1dbe6c3a8713b97deb2e150f` |
| CrispASR CUDA runtime archive | `832af6218508ac52fc71ac5653c433786bdfae81f775e2c77bfefd05df6f255b` |
| Silero VAD revision | `9ffd54a1e1ee413ddf265af9913beaf518d1639b` |
| ECAPA language-ID revision | `95fb0613bf78c6e48305fccd9ce023ac15f0b5a6` |
| `FluidInference/fleurs` | `8944693da251acbaf2f9686bddc4fedce8bd2edd` |
| `FluidInference/fleurs-full` | `1cca811bb8ea4d370345f108f00518167040282c` |
| Sample manifest SHA-256 | `36e8a55f93116011e1cdd878fec4bd65c239dbe5357cd58e4a4d6fe42a185c46` |

Model artifacts:

| Variant | Bytes | Decimal size | SHA-256 |
|---|---:|---:|---|
| Q4_K | 1,510,362,752 | 1.51 GB | `2931fc0ac6d6708eef5389aadf1ebd5eec7b8e764bac385be585e910c0e7b410` |
| Q5_0 | 1,738,722,944 | 1.74 GB | `a09696c5cc2ed5052bf290c4f2beb35abc69c0d6986842042d92bebb22c9184e` |
| Q6_K | 1,981,355,648 | 1.98 GB | `0ad2634e0ba34efa38a47d4fd4cf34d7a2d738d8486d83b8d5a178f823109c52` |
| Q8_0 | 2,423,803,520 | 2.42 GB | `c8620cb182a7c04e311e6c24e478b94f7ecd7f1b5230bf39fffa8daf94644f51` |

### Test design

- Exactly the first 50 pinned FLEURS test clips were used for each of the
  model's 14 supported languages.
- 14 languages × 50 clips × 4 quantizations = **2,800 speech
  transcriptions**.
- Every model received the same clips in the same order and the correct
  explicit language hint.
- One persistent CrispASR server process was used per quantization. Model load
  and server startup were recorded separately and were not included in warm
  inference throughput.
- The runner called the local CrispASR transcription endpoint directly. The
  TypeWhisper API was used only to verify/prepare model assets.
- TypeWhisper dictionary terms, dictionary corrections, Workflows, and number
  normalization were not applied. Results are raw CrispASR output.
- WER is the primary metric except for Japanese, Mandarin Chinese, and Korean,
  where CER is used because whitespace-based WER is not comparable.
- Text normalization is Unicode NFKC + case folding + punctuation/symbol
  removal.
- RTFx is total audio duration divided by client wall time around the direct
  localhost inference request. Higher is faster.
- Peak process working set was read from the exact persistent CrispASR PID.
  It is **not** a GPU-VRAM measurement.
- Checkpoints were written atomically every 10 clips and can be resumed.
- The same 1-second silence, 5-second silence, and 5-second low-white-noise
  PCM16/16 kHz inputs were tested for every variant.

The corpus contains 8,205.46 seconds (2 h 16 min 45 s) of audio per model.
This is a controlled read-speech comparison, not a claim that the values will
exactly match conversational microphone dictation.

## Aggregate results

| Variant | Model size | Non-CJK macro WER | CJK macro CER | Weighted RTFx | Startup | Peak working set |
|---|---:|---:|---:|---:|---:|---:|
| Q4_K | 1.51 GB | **7.67%** | 8.30% | 32.27x | 3.39 s | **1.66 GiB** |
| Q5_0 | 1.74 GB | 7.90% | 8.29% | **32.69x** | **2.79 s** | 1.87 GiB |
| Q6_K | 1.98 GB | 7.89% | **8.27%** | 32.64x | 3.29 s | 2.10 GiB |
| Q8_0 | 2.42 GB | 7.93% | **8.27%** | 31.78x | 3.80 s | 2.51 GiB |

The full 700-clip inference totals were:

| Variant | Audio | Inference time | RTFx |
|---|---:|---:|---:|
| Q4_K | 8,205.46 s | 254.25 s | 32.27x |
| Q5_0 | 8,205.46 s | 251.04 s | 32.69x |
| Q6_K | 8,205.46 s | 251.38 s | 32.64x |
| Q8_0 | 8,205.46 s | 258.18 s | 31.78x |

## Per-language accuracy

Each cell contains the primary metric for 50 clips. Lower is better.

| Language | Metric | Q4_K | Q5_0 | Q6_K | Q8_0 |
|---|---|---:|---:|---:|---:|
| German | WER | **4.00%** | 4.33% | 4.25% | 4.42% |
| English (US) | WER | 6.13% | 6.78% | 6.04% | **5.95%** |
| French | WER | 4.90% | 5.42% | **4.82%** | 5.12% |
| Spanish (Latin America) | WER | **2.94%** | **2.94%** | 3.02% | 3.02% |
| Italian | WER | **2.71%** | 3.39% | 3.31% | 3.61% |
| Portuguese (Brazil) | WER | **9.73%** | 9.97% | 10.13% | 10.21% |
| Greek | WER | **9.62%** | 10.18% | 10.34% | 10.10% |
| Dutch | WER | 7.86% | **7.59%** | 7.86% | 7.94% |
| Polish | WER | **7.13%** | 7.24% | **7.13%** | 7.24% |
| Japanese | CER | 8.46% | **8.26%** | 8.38% | 8.38% |
| Mandarin Chinese | CER | **12.42%** | 12.59% | **12.42%** | **12.42%** |
| Korean | CER | **4.02%** | **4.02%** | **4.02%** | **4.02%** |
| Vietnamese | WER | 8.33% | **7.84%** | 8.19% | 8.19% |
| Arabic | WER | **20.99%** | 21.21% | 21.76% | 21.43% |

The largest non-CJK macro-WER spread is 0.26 percentage points. The CJK
macro-CER spread is 0.03 percentage points. The per-language winner varies, so
there is no evidence here that increasing quantization size consistently
improves accuracy.

## Completion and negative tests

- Successful speech clips: **2,800 / 2,800**
- Failed speech clips: **0**
- Empty speech results: **0**
- Silence/noise hallucinations: **0 / 12**

| Variant | 1 s silence | 5 s silence | 5 s low white noise |
|---|---:|---:|---:|
| Q4_K | Empty (0.016 s) | Empty (0.040 s) | Empty (0.064 s) |
| Q5_0 | Empty (0.016 s) | Empty (0.059 s) | Empty (0.062 s) |
| Q6_K | Empty (0.016 s) | Empty (0.049 s) | Empty (0.061 s) |
| Q8_0 | Empty (0.015 s) | Empty (0.053 s) | Empty (0.061 s) |

The integrated Windows plugin launches CrispASR with mandatory Silero VAD, so
these results should not be generalized to a VAD-free Cohere pipeline.

### Prerequisites

1. Windows x64 with TypeWhisper and Cohere Transcribe (Local) plugin 1.0.1.
2. Enable the TypeWhisper local API (default URL:
   `http://127.0.0.1:8978`).
3. Download the variants to compare and load the intended acceleration backend
   once so its CrispASR runtime is present.
4. Python 3.10 or newer. The runner uses only the standard library.
5. Allow roughly 8 GB for all four model files, plus the selected runtime and
   FLEURS cache.

Checkout the exact benchmark source:

```powershell
git clone https://github.com/TypeWhisper/typewhisper-win.git
Set-Location typewhisper-win
git checkout 5e69ab6d4ec4b7c71c625f2c17bb062f7f408852
```

Optional authenticated Hugging Face downloads:

```powershell
$env:HF_TOKEN = "hf_..."
```

Run the CUDA benchmark:

```powershell
py -3 eng/benchmark_cohere_quantizations.py `
  --backend nvidia-cuda `
  --stop-typewhisper-sidecar
```

The runner auto-detects a complete Store, desktop, or development asset root
when only one exists. If multiple complete installations exist, pass the root
explicitly:

```powershell
py -3 eng/benchmark_cohere_quantizations.py `
  --asset-root "C:\path\to\PluginData\com.typewhisper.cohere-transcribe" `
  --backend nvidia-cuda `
  --stop-typewhisper-sidecar
```

The stop flag terminates only a `crispasr.exe` whose executable is below the
selected Cohere asset root; the TypeWhisper app remains open. The runner then
owns one persistent CrispASR process at a time and shuts it down after each
model.

Resume an interrupted compatible checkpoint:

```powershell
py -3 eng/benchmark_cohere_quantizations.py `
  --backend nvidia-cuda `
  --stop-typewhisper-sidecar `
  --resume
```

Use `--backend amd-vulkan` or `--backend cpu` for the other Windows runtimes.
Use `--models` and `--languages` for smaller diagnostic runs.

Default outputs:

- `artifacts/cohere-quantization-benchmark/direct-results.json`
- `artifacts/cohere-quantization-benchmark/direct-results.md`

The JSON output from the run summarized above has SHA-256:

`9af597b1335169fb2358d2ece1a72fcfdfcdc258fe5120377b9cd42856641f8a`

## Sharing another result

Please include:

- CPU, GPU, GPU memory, system memory, OS build, and driver;
- backend and thread count;
- TypeWhisper/plugin version and runner commit;
- exact command;
- sample manifest SHA-256;
- the aggregate table from `direct-results.md`;
- successful/failed clip and negative-test counts;
- whether the run was uninterrupted or resumed.

Do not compare accuracy rows if the sample manifest differs. Performance
comparisons should also keep the language set, sample count, backend, and thread
count explicit.

## Limitations

- One Windows machine and one CUDA backend were measured.
- FLEURS is read speech, not live TypeWhisper microphone dictation.
- Fifty clips per language are useful for a controlled comparison but do not
  provide narrow confidence intervals for small quantization differences.
- Peak working set is process memory, not GPU VRAM.
- Startup is separated from warm inference, so these RTFx values are not
  end-to-end first-dictation latency.
- TypeWhisper post-processing was intentionally excluded to isolate the local
  model/runtime behavior.
- This issue compares Cohere quantizations with each other. For the broader
  Cohere-versus-Parakeet product analysis on Apple Silicon, see
  TypeWhisper/typewhisper-mac#1031.

## Sources

- [Cohere Transcribe 03-2026 model card](https://huggingface.co/CohereLabs/cohere-transcribe-03-2026)
- [Cohere Transcribe GGUF conversion](https://huggingface.co/cstr/cohere-transcribe-03-2026-GGUF)
- [Windows implementation and selectable variants](https://github.com/TypeWhisper/typewhisper-win/pull/338)
- [Related Apple Silicon Cohere/Parakeet benchmark](https://github.com/TypeWhisper/typewhisper-mac/issues/1031)

### PR DESCRIPTION
**Pull request by SeoFood**
## Summary
- start CrispASR outside the inherited MSIX desktop runtime so native DLLs resolve from its runtime directory
- translate redirected LocalAppData paths to the physical Store package data path for the unpackaged child process
- offer Q4_K, Q5_0, Q6_K, and Q8_0 as separate, individually downloadable Cohere models while sharing auxiliary models and the selected runtime
- add a pinned, resumable FLEURS benchmark that keeps one persistent CrispASR process per quantization
- recover automatically when the app-owned CrispASR sidecar exits unexpectedly while a model is marked as loaded
- bump the Cohere Transcribe plugin to 1.0.1 and cover the launch, path, model-selection, and sidecar-recovery behavior with regression tests

## Verification
- 46 focused Cohere Transcribe tests pass
- Release plugin build passes with 0 warnings and 0 errors
- Python compilation, formatting, and diff checks pass
- installed the exact plugin build into TypeWhisper 1.0.6.0 from Microsoft Store and confirmed all four model variants are listed separately
- reused the existing Q5_0 download and completed an end-to-end CUDA transcription without a DLL error dialog
- terminated the active Q5_0 CUDA sidecar and confirmed the same request restarted it under a new process without restarting TypeWhisper
- completed the reproducible CUDA benchmark: 2,800/2,800 speech clips, 0 failures, and 0/12 silence/noise hallucinations
- confirmed the benchmark runner uses a persistent process per model and the Store app remains open when its sidecar is released
- confirmed the CrispASR process tree terminates with the Store host

Benchmark methodology and results: #339

Closes #339

### CODE REVIEW

File: plugins/TypeWhisper.Plugin.CohereTranscribe/CrispAsrServer.cs [hunk 1]
```diff
@@ -202,41 +207,128 @@ internal static ProcessStartInfo BuildStartInfo(
...
+                StringComparison.OrdinalIgnoreCase);
+    }
+
+    private static string? TryGetCurrentPackageFamilyName()
+    {
+        if (!OperatingSystem.IsWindows())
+            return null;
+
+        uint length = 0;
+        var result = GetCurrentPackageFamilyName(ref length, null);
```
[the remark this answers is not in the document]
Comment (SeoFood) [reply plugins/TypeWhisper.Plugin.CohereTranscribe/CrispAsrServer.cs]: Replaced the package identity P/Invoke with the managed Windows.Storage.ApplicationData API and revalidated the change in the installed Store app.

File: plugins/TypeWhisper.Plugin.CohereTranscribe/CrispAsrServer.cs [hunk 2]
```diff
@@ -202,41 +207,128 @@ internal static ProcessStartInfo BuildStartInfo(
...
+        if (result != ErrorInsufficientBuffer || length == 0)
+            return null;
+
+        var familyName = new StringBuilder((int)length);
+        result = GetCurrentPackageFamilyName(ref length, familyName);
```
[the remark this answers is not in the document]
Comment (SeoFood) [reply plugins/TypeWhisper.Plugin.CohereTranscribe/CrispAsrServer.cs]: Replaced the package identity P/Invoke with the managed Windows.Storage.ApplicationData API and revalidated the change in the installed Store app.

File: plugins/TypeWhisper.Plugin.CohereTranscribe/CrispAsrServer.cs [hunk 3]
```diff
@@ -316,9 +408,8 @@ private static string GetBackendName(CrispAsrBackend backend) =>
             _ => throw new ArgumentOutOfRangeException(nameof(backend), backend, null)
         };
 
-    private static void AddArgument(ProcessStartInfo startInfo, params string[] values)
-    {
-        foreach (var value in values)
-            startInfo.ArgumentList.Add(value);
-    }
+    [DllImport("kernel32.dll", CharSet = CharSet.Unicode)]
+    private static extern int GetCurrentPackageFamilyName(
```
[the remark this answers is not in the document]
Comment (SeoFood) [reply plugins/TypeWhisper.Plugin.CohereTranscribe/CrispAsrServer.cs]: Removed the unmanaged declaration entirely. Store path resolution now uses ApplicationData.Current.LocalCacheFolder.

**Comment by SeoFood** [comment]
Addressed the final review findings in 7ff64e3: null-safe negative-test reporting, per-sample negative-test failures, independent artifact pins, selected-model path coverage, startup exception chaining, numeric argument validation, explicit loop bindings, and the cmd.exe/runtime-path assertions and documentation. Revalidated 44 focused tests, formatting, parser edge cases, report error handling, and a direct CUDA smoke run.

**Comment by SeoFood** [comment]
Live Store follow-up: releasing the benchmark sidecar exposed a stale loaded-model state when CrispASR exited unexpectedly. Commit 323c19e adds a one-time sidecar restart and transcription retry. I validated it in the installed TypeWhisper 1.0.6.0 Store app by terminating the active Q5_0 CUDA sidecar (PID 661368); the same selected-model request succeeded, CrispASR restarted as PID 673220, and TypeWhisper itself was not restarted. The focused Cohere suite now passes 46/46 tests, including both pre-request and mid-request sidecar exits.

**Comment by SeoFood** [comment]
Released as [Cohere Transcribe plugin v1.0.1](https://github.com/TypeWhisper/typewhisper-win/releases/tag/plugin-cohere-transcribe-v1.0.1). The public plugin registry now serves v1.0.1, and the GitHub Release asset and GitHub Pages download both match SHA-256 \dc61b53cece5d2b287012ca551b5789227b24d75610ef7134a3163642e6b9538\ (42,614 bytes).
